### DISCUSSION TITLE
Is there any way to run a Rust function on Dart VM shutdown?

### DISCUSSION THREAD

**Discussion by temeddix** [sunshine-protocol/allo-isolate#51] [Q&A] [answered]
I have some resources such as databases on the Rust side. When Dart VM gets shut down, perhaps because the user closed the Flutter app, the Rust finalizer functions that destroy Rust resources should be run.

Is this possible with `allo-isolate`, or can be made possible with a PR?

**Comment by shekohex** [comment, marked as answer]
If there a way to do some actions/[lifecycle-hooks](https://api.flutter.dev/flutter/widgets/AppLifecycleListener-class.html) when the user closes the flutter application then you can just call your rust function from there, I do not see any reason why that wouldn't work.

**Comment by temeddix** [comment]
I see, that would be the reasonable way to do handle shutdown. Thank you :)

**Comment by temeddix** [comment]
Unfortunately, Flutter's lifecycle hooks only work on mobile platforms and not on the desktop. I've been testing lifecycle hooks on desktop platforms, and they don't do anything related to `onDetach`, that should be fired on app exit.

- https://stackoverflow.com/questions/67881349/flutter-desktop-cannot-respond-to-lifecycle-events
- https://stackoverflow.com/questions/71458825/flutter-widgetsbindingobservers-didchangeapplifecyclestate-not-being-called-on
- https://github.com/flutter/flutter/issues/28879

`AppLifecycleListener`'s `onExitRequested` is mainly for **deciding whether to exit**, so it's not suitable for our case.

**Reply by shekohex** [reply]
Yeah, I assumed you are talking about mobile devices, since you did not state otherwise. But thanks for the information, may it help someone else searching for the same issue.
